crates/bevy_ecs/macros/src/fetch.rs

```diff
@@ -0,0 +1,622 @@
+    let mut field_idents = Vec::new();
+    let mut field_types = Vec::new();
+    let mut query_types = Vec::new();
+    let mut fetch_init_types = Vec::new();
+
+    for field in fields.iter() {
+        let (world_lifetime, fetch_lifetime) = match (&world_lifetime, &fetch_lifetime) {
+            (GenericParam::Lifetime(world), GenericParam::Lifetime(fetch)) => {
+                (&world.lifetime, &fetch.lifetime)
+            }
+            _ => unreachable!(),
+        };
+        let WorldQueryFieldTypeInfo {
+            query_type,
+            fetch_init_type: init_type,
+            is_phantom_data,
+        } = read_world_query_field_type_info(&field.ty, world_lifetime, fetch_lifetime);
+
+        let field_ident = field.ident.as_ref().unwrap().clone();
+        if is_phantom_data {
+            phantom_field_idents.push(field_ident.clone());
+            phantom_field_types.push(field.ty.clone());
+        } else {
+            field_idents.push(field_ident.clone());
+            field_types.push(field.ty.clone());
+            query_types.push(query_type);
+            fetch_init_types.push(init_type);
+        }
+    }
+
+    FetchImplTokens {
+        struct_name,
+        struct_name_read_only,
+        fetch_struct_name,
+        state_struct_name,
+        read_only_fetch_struct_name,
+        fetch_trait_punctuated_lifetimes,
+        impl_generics,
+        ty_generics,
+        where_clause,
+        has_mutable_attr,
+        has_world_lifetime,
+        world_lifetime,
+        state_lifetime,
+        fetch_lifetime,
+        phantom_field_idents,
+        phantom_field_types,
+        field_idents,
+        field_types,
+        query_types,
+        fetch_init_types,
+    }
+}
+
+struct WorldQueryFieldTypeInfo {
+    /// We convert `Mut<T>` to `&mut T` (because this is the type that implements `WorldQuery`)
+    /// and store it here.
+    query_type: Type,
+    /// The same as `query_type` but with `'fetch` lifetime.
+    fetch_init_type: Type,
+    is_phantom_data: bool,
+}
+
+fn read_world_query_field_type_info(
+    ty: &Type,
+    world_lifetime: &Lifetime,
+    fetch_lifetime: &Lifetime,
+) -> WorldQueryFieldTypeInfo {
+    let path = bevy_ecs_path();
+
+    let query_type = parse_quote!(<#ty as #path::query::FetchedItem>::Query);
+    let mut fetch_init_type = parse_quote!(<#ty as #path::query::FetchedItem>::Query);
+
+    let is_phantom_data = match ty {
+        Type::Path(path) => {
+            if let Some(segment) = path.path.segments.last() {
+                segment.ident == "PhantomData"
+            } else {
+                false
+            }
+        }
+        _ => false,
+    };
+
+    replace_lifetime_for_type(&mut fetch_init_type, world_lifetime, fetch_lifetime);
+
+    WorldQueryFieldTypeInfo {
+        query_type,
+        fetch_init_type,
+        is_phantom_data,
+    }
+}
+
+fn replace_lifetime_for_type(ty: &mut Type, world_lifetime: &Lifetime, fetch_lifetime: &Lifetime) {
```

Comment:
I think this function deserves another copy of this comment.
```suggestion
// Fetch's HRTBs require substituting world lifetime with an additional one to make the
// implementation compile. I don't fully understand why this works though. If anyone's curious
// enough to try to find a better work around, I'll leave playground links here:
// - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=da5e260a5c2f3e774142d60a199e854a (this fails)
// - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=802517bb3d8f83c45ee8c0be360bb250 (this compiles)
fn replace_lifetime_for_type(ty: &mut Type, world_lifetime: &Lifetime, fetch_lifetime: &Lifetime) {
```